### PR TITLE
Impro 931 adjust weights spatially cli changes

### PR DESCRIPTION
**Pull request by fionaRust**
Addresses #692

Changes to CLI to use new spatial weights option. When writing new tests for this spatial weights functionality I found that there were several problems with the dimensions on weights cubes, so there are extra changes to fix these things. 

Testing:
 - [ ] Ran tests and they passed OK
 - [ ] Added new tests for the new feature(s)

### CODE REVIEW

File: lib/improver/blending/spatial_weights.py [hunk 1]
```diff
@@ -255,6 +255,80 @@ def normalised_masked_weights(weights_cube, blend_coord):
             result.append(weight_slice.copy(data=normalised_data))
         return result.merge_cube()
 
+    @staticmethod
+    def create_template_slice(cube_to_collapse, blend_coord):
```
Comment (fionaRust) [inline lib/improver/blending/spatial_weights.py]: This function was created to change the dimensions in the output cube for the spatially varying weights. Before it returned a weights cube with the same dimensions as the cube_to_collapse. Now it returns a cube with dimensions of blend_coord, y x.
This was done to make the weighted blending code work better, and is a better representation of how the weights vary. (Before it may return a weights cube with dimensions of blend_coord, threshold, y, x even though the weights didn't vary along the threshold dimension.)
I haven't yet written unit tests for this, but ideas on what unit tests are needed are on the ticket.

File: lib/improver/tests/blending/spatial_weights/test_SpatiallyVaryingWeightsFromMask.py [hunk 2]
```diff
@@ -653,18 +653,12 @@ def test_none_masked(self, warning_list=None):
...
-                                    [0.2, 0.2, 0.2]]],
-                                  [[[0.5, 0.5, 0.5],
-                                    [0.5, 0.5, 0.5]],
-                                   [[0.5, 0.5, 0.5],
-                                    [0.5, 0.5, 0.5]]],
-                                  [[[0.3, 0.3, 0.3],
-                                    [0.3, 0.3, 0.3]],
-                                   [[0.3, 0.3, 0.3],
-                                    [0.3, 0.3, 0.3]]]],
+        expected_data = np.array([[[0.2, 0.2, 0.2],
```
Comment (fionaRust) [inline lib/improver/tests/blending/spatial_weights/test_SpatiallyVaryingWeightsFromMask.py]: All these unit tests have been updated to remove any extra dimensions (like threshold) that the weights don't vary across.

File: lib/improver/blending/weighted_blend.py [hunk 3]
```diff
@@ -573,12 +568,17 @@ def shape_weights(self, cube, weights):
             # Multi-dimensional weights array provided that matches cube.
             weights_array = weights.data.astype(np.float32)
         else:
-            # 1D array of weights across the blending coordinate.
-            coord_dim_thres = cube.coord_dims(self.coord)
+            # Map array of weights to shape of cube to collapse.
```
Comment (fionaRust) [inline lib/improver/blending/weighted_blend.py]: Before this function could only cope with weights with 1D or weights that exactly matched the dims on the cube to collapse. I've updated the broadcasting to be a bit more general, so one unit test now passes when it failed (we need to add some expected data to this test), but we probably need a few more unit tests to test this broadcasting properly.
See details on the ticket.

File: lib/improver/blending/weighted_blend.py [hunk 4]
```diff
@@ -870,51 +871,28 @@ def process(self, cube, weights=None):
         # Check to see if the data is percentile data
         perc_coord = self.check_percentile_coord(cube)
 
-        # Create slices over the threshold coordinate
-        try:
```
Comment (fionaRust) [inline lib/improver/blending/weighted_blend.py]: I also found that this slicing over threshold was unnecessary and complicated things when you had 3D weights, so I removed it.

Comment (bayliffe) [reply lib/improver/blending/weighted_blend.py]: Huzzah, simplification!

File: lib/improver/tests/blending/spatial_weights/test_SpatiallyVaryingWeightsFromMask.py (hunk 2 above)
Comment (bayliffe) [reply lib/improver/tests/blending/spatial_weights/test_SpatiallyVaryingWeightsFromMask.py]: Otherwise unchanged. Seem fine.

File: lib/improver/tests/blending/weighted_blend/test_WeightedBlendAcrossWholeDimension.py [hunk 5]
```diff
@@ -321,6 +321,7 @@ def test_incompatible_weights_and_data_cubes(self):
         msg = "Weights cube is not a compatible shape with the data cube"
 
         with self.assertRaisesRegex(ValueError, msg):
+
```
Comment (bayliffe) [inline lib/improver/tests/blending/weighted_blend/test_WeightedBlendAcrossWholeDimension.py]: I enjoy the new found spaciousness provided by this line. I very much felt the feng shui was all wrong before, but now the qi of this test file seems to flow all the better 😄

File: tests/improver-weighted-blending/19-spatial_model_blending.bats [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+# POSSIBILITY OF SUCH DAMAGE.
+
+. $IMPROVER_DIR/tests/lib/utils
+
+@test "weighted-blending model blending" {
+  improver_check_skip_acceptance
+  KGO="weighted_blending/spatial_weights/nowcast_blending/model_kgo.nc"
+
+  # Run weighted blending with linear weights for two input files and check it
+  # passes.
```
Comment (bayliffe) [inline tests/improver-weighted-blending/19-spatial_model_blending.bats]: This comment could be more relevant.

File: tests/improver-weighted-blending/20-nowcast_cycle_blending_no_fuzzy.bats [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+# POSSIBILITY OF SUCH DAMAGE.
+
+. $IMPROVER_DIR/tests/lib/utils
+
+@test "weighted-blending model blending" {
+  improver_check_skip_acceptance
+  KGO="weighted_blending/spatial_weights/nowcast_blending/kgo_cycle_no_fuzzy.nc"
+
+  # Run weighted blending with linear weights for two input files and check it
+  # passes.
```
Comment (bayliffe) [inline tests/improver-weighted-blending/20-nowcast_cycle_blending_no_fuzzy.bats]: And again.

File: tests/improver-weighted-blending/21-spatial_model_blending_no_fuzzy.bats [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+# POSSIBILITY OF SUCH DAMAGE.
+
+. $IMPROVER_DIR/tests/lib/utils
+
+@test "weighted-blending model blending" {
+  improver_check_skip_acceptance
+  KGO="weighted_blending/spatial_weights/nowcast_blending/model_kgo_no_fuzzy.nc"
+
+  # Run weighted blending with linear weights for two input files and check it
+  # passes.
```
Comment (bayliffe) [inline tests/improver-weighted-blending/21-spatial_model_blending_no_fuzzy.bats]: etc.

File: tests/improver-weighted-blending/20-nowcast_cycle_blending_no_fuzzy.bats [hunk 9]
```diff
@@ -0,0 +1,53 @@
...
+# CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+# SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+# INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+# CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+# ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+# POSSIBILITY OF SUCH DAMAGE.
+
+. $IMPROVER_DIR/tests/lib/utils
+
+@test "weighted-blending model blending" {
```
Comment (bayliffe) [inline tests/improver-weighted-blending/20-nowcast_cycle_blending_no_fuzzy.bats]: This needs to be updated for this test.

File: tests/improver-weighted-blending/21-spatial_model_blending_no_fuzzy.bats [hunk 10]
```diff
@@ -0,0 +1,52 @@
...
+# CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+# SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+# INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+# CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+# ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+# POSSIBILITY OF SUCH DAMAGE.
+
+. $IMPROVER_DIR/tests/lib/utils
+
+@test "weighted-blending model blending" {
```
Comment (bayliffe) [inline tests/improver-weighted-blending/21-spatial_model_blending_no_fuzzy.bats]: This needs to be updated for this test.

File: tests/improver-weighted-blending/22-spatial_percentile_blending.bats [hunk 11]
```diff
@@ -0,0 +1,53 @@
...
+# CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+# SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+# INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+# CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+# ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+# POSSIBILITY OF SUCH DAMAGE.
+
+. $IMPROVER_DIR/tests/lib/utils
+
+@test "weighted-blending model blending" {
```
Comment (bayliffe) [inline tests/improver-weighted-blending/22-spatial_percentile_blending.bats]: This needs to be updated for this test.

File: bin/improver-weighted-blending [hunk 12]
```diff
@@ -182,6 +172,18 @@ def main():
                              'multi-model blends. Default assumes Met Office '
                              'model metadata. Must be present on all input '
                              'files if blending over models.')
+    parser.add_argument('--spatial_weights_from_mask',
+                        action='store_true', default=False,
+                        help='If set this option will result in the generation'
+                             ' of spatially varying weights based on the'
+                             ' masks of the data we are blending. The'
+                             ' one dimension weights are first calculated '
```
[the remark this answers is not in the document]
Comment (fionaRust) [reply bin/improver-weighted-blending]: fixed

File: lib/improver/blending/spatial_weights.py [hunk 13]
```diff
@@ -255,6 +255,80 @@ def normalised_masked_weights(weights_cube, blend_coord):
...
+            ValueError: if the blend coordinate is associated with more than
+                        one dimension on the cube to collapse
+            ValueError: if the mask on cube_to_collapse varies along a
+                        dimension other than the dimension associated with
+                        blend_coord.
+        """
+        # Takes slices over x, y coord_to_collapse
+        # Find blend dim coord associated with blend coord, sometimes we name a
+        # blend_coord that is a aux_coord associated with a dim coord rather
+        # than using name of the dim_coord itself.
```
[the remark this answers is not in the document]
Comment (fionaRust) [reply lib/improver/blending/spatial_weights.py]: fixed

File: lib/improver/blending/spatial_weights.py [hunk 14]
```diff
@@ -266,7 +340,9 @@ def process(self, cube_to_collapse, one_dimensional_weights_cube,
             cube_to_collapse (iris.cube.Cube):
                 The cube that will be collapsed along the blend_coord
                 using the spatial weights generated using this plugin. Must
-                be masked where there is invalid data.
+                be masked where there is invalid data. The mask must only
+                vary along the blend_coord, an not along any other dimensions
+                on the cube.
```
[the remark this answers is not in the document]
Comment (fionaRust) [reply lib/improver/blending/spatial_weights.py]: Fixed

File: tests/improver-weighted-blending/19-spatial_model_blending.bats (hunk 6 above)
Comment (fionaRust) [reply tests/improver-weighted-blending/19-spatial_model_blending.bats]: Fixed

File: tests/improver-weighted-blending/20-nowcast_cycle_blending_no_fuzzy.bats (hunk 9 above)
Comment (fionaRust) [reply tests/improver-weighted-blending/20-nowcast_cycle_blending_no_fuzzy.bats]: Fixed

File: tests/improver-weighted-blending/20-nowcast_cycle_blending_no_fuzzy.bats (hunk 7 above)
Comment (fionaRust) [reply tests/improver-weighted-blending/20-nowcast_cycle_blending_no_fuzzy.bats]: fixed

File: tests/improver-weighted-blending/21-spatial_model_blending_no_fuzzy.bats (hunk 10 above)
Comment (fionaRust) [reply tests/improver-weighted-blending/21-spatial_model_blending_no_fuzzy.bats]: Fixed

File: tests/improver-weighted-blending/21-spatial_model_blending_no_fuzzy.bats (hunk 8 above)
Comment (fionaRust) [reply tests/improver-weighted-blending/21-spatial_model_blending_no_fuzzy.bats]: fixed

**Comment by fionaRust** [comment]
I've now added all the new unit tests described on the ticket, and responded to the first set of Ben's comments.

File: lib/improver/blending/weighted_blend.py (hunk 4 above)
Comment (cgsandford) [reply lib/improver/blending/weighted_blend.py]: :+1:

File: lib/improver/tests/blending/weights/test_WeightsUtilities.py [hunk 15]
```diff
@@ -278,18 +278,21 @@ def test_basic_weights(self):
...
-
+        expected_coord_names = ["time", "forecast_reference_time",
+                                "forecast_period"]
         self.assertIsInstance(result, iris.cube.Cube)
         self.assertEqual(result.name(), 'weights')
         self.assertFalse(result.attributes)
         self.assertArrayEqual(result.data, weights)
         self.assertEqual(result.coords(dim_coords=True)[0].name(),
-                         blending_coord)
+                         "time")
```
Comment (cgsandford) [inline lib/improver/tests/blending/weights/test_WeightsUtilities.py]: Why these changes?

File: bin/improver-weighted-blending [hunk 16]
```diff
@@ -197,6 +199,22 @@ def main():
     parser.add_argument('output_filepath', metavar='OUTPUT_FILE',
                         help='The output path for the processed NetCDF.')
 
+    spatial = parser.add_argument_group(
+        'Spatial weights from mask options',
+        'Options for calculating the spatial weights using the '
+        'SpatiallyVaryingWeightsFromMask plugin.')
+    spatial.add_argument('--fuzzy_length', metavar='FUZZY_LENGTH', type=float,
+                         default=10,
```
Comment (cgsandford) [inline bin/improver-weighted-blending]: I would prefer to see this as a distance in metres or kilometres please, for consistency with other CLIs.

File: lib/improver/tests/blending/weights/test_WeightsUtilities.py (hunk 15 above)
Comment (fionaRust) [reply lib/improver/tests/blending/weights/test_WeightsUtilities.py]: This was trying to test the change I made to the build_weights_cube function to keep all the coordinates associated with the blending coord, not just the dim coord, so in this case we're keeping both the forecast_period and time coordinates, not just the forecast_period coordinate

File: bin/improver-weighted-blending (hunk 16 above)
Comment (fionaRust) [reply bin/improver-weighted-blending]: I can change this, but I think it would need a plugin change to convert the distance to a number of grid points.

File: lib/improver/tests/blending/spatial_weights/test_SpatiallyVaryingWeightsFromMask.py [hunk 17]
```diff
@@ -590,6 +591,110 @@ def test_weights_sum_to_zero(self):
...
+        self.cube_to_collapse.add_aux_coord(altitudes_coord, data_dims=(2, 3))
+        message = ("Blend coordinate must only be across one dimension.")
+        with self.assertRaisesRegex(ValueError, message):
+            self.plugin.create_template_slice(
+                self.cube_to_collapse, "surface_altitude")
+
+    def test_varying_mask_fail(self):
+        """Test error is raised when mask varies along collapsing dim"""
+        # Check fails when blending along threshold coordinate, as mask
+        # varies along this coodinate.
```
[the remark this answers is not in the document]
Comment (fionaRust) [reply lib/improver/tests/blending/spatial_weights/test_SpatiallyVaryingWeightsFromMask.py]: Fixed

File: bin/improver-weighted-blending (hunk 16 above)
Comment (cgsandford) [reply bin/improver-weighted-blending]: You could call the spatial utilities function from the CLI or, yes, from the plugin if needed.

Did we have this conversation about the plugin already?

File: lib/improver/blending/weights.py [hunk 18]
```diff
@@ -285,13 +285,14 @@ def build_weights_cube(cube, weights, blending_coord):
         except ValueError:
             weights_cube = iris.util.new_axis(cube, blending_coord)
             weights_cube = next(weights_cube.slices(blending_coord))
-
         weights_cube.attributes = None
+        # Find dim associated with blending_coord and don't remove any coords
```
[the remark this answers is not in the document]
Comment (cgsandford) [reply lib/improver/blending/weights.py]: Agreed - can you give an example of when we were not doing that?  (I don't think we have had issues with losing "time" or "forecast_period" when cycle blending, for example.)

I think this change is fine in principle but am concerned it might have unexpected effects, eg with multi-model blending.

File: bin/improver-weighted-blending (hunk 16 above)
Comment (fionaRust) [reply bin/improver-weighted-blending]: Yes we did. See this discussion: https://github.com/metoppv/improver/pull/753#discussion_r238225965

File: lib/improver/blending/weights.py (hunk 18 above)
[the remark this answers is not in the document]
Comment (fionaRust) [reply lib/improver/blending/weights.py]: At some point in during this ticket I had a problem with anonymous dimensions appearing on my weights cube. I think I still have this problem with one test now (I think it's with model blending not cycle blending). I'm taking a look at it now, and I'll put up the details on the ticket.

[the remark this answers is not in the document]
Comment (fionaRust) [reply lib/improver/blending/weights.py]: The problem with this is when we do model blending. Without this change we get a weights cube of:
```
weights: weights / (1)                       (-- : 2)
     Auxiliary coordinates:
          model_configuration           x
     Scalar coordinates:
          forecast_period: 18000 seconds
          forecast_reference_time: 2017-12-08 04:00:00
          time: 2017-12-08 09:00:00
```
and with this change we get
```
weights: weights / (1)                       (model_id: 2)
     Dimension coordinates:
          model_id                           x
     Auxiliary coordinates:
          model_configuration                x
     Scalar coordinates:
          forecast_period: 18000 seconds
          forecast_reference_time: 2017-12-08 04:00:00
          time: 2017-12-08 09:00:00
```
The second case makes it much easier to match coordinates in the weights cube and the cube we are collapsing.
Bats test 12 and 17 and affected by this change and they currently fail without it.

File: lib/improver/blending/weighted_blend.py [hunk 19]
```diff
@@ -560,26 +560,30 @@ def shape_weights(self, cube, weights):
...
-            coord_dim_thres = cube.coord_dims(self.coord)
+            # Map array of weights to shape of cube to collapse.
+            dim_map = []
+            dim_coords = [coord.name() for coord in weights.dim_coords]
+            # Loop through dim coords in weights cube and find the dim the
+            # coord relates to in the cube we are collapsing.
+            for dim_coord in dim_coords:
+                try:
+                    dim_map.append(cube.coord_dims(dim_coord)[0])
+                except CoordinateNotFoundError as E:
```
[the remark this answers is not in the document]
Comment (fionaRust) [reply lib/improver/blending/weighted_blend.py]: Fixed

File: lib/improver/tests/blending/weights/test_WeightsUtilities.py (hunk 15 above)
Comment (cgsandford) [reply lib/improver/tests/blending/weights/test_WeightsUtilities.py]: OK.  I think we do want to keep a test that blends across forecast period, as this is a very widely used case (and of course blends across an aux_coord).  Should this not also retain the time and forecast reference time coordinates (although only time shares a dimension with forecast period)?  I could see the case for a new test blending across time, but would rather not lose the existing one.

File: lib/improver/blending/weights.py (hunk 18 above)
[the remark this answers is not in the document]
Comment (cgsandford) [reply lib/improver/blending/weights.py]: OK, it's clear this change is an improvement then - I'm happy with this.

File: bin/improver-weighted-blending (hunk 16 above)
Comment (cgsandford) [reply bin/improver-weighted-blending]: Ah I see.  I think I'm still OK with that decision at the plugin level, but it doesn't sit right with me for a CLI to be taking arguments in subjective units.  Perhaps you could do a float conversion in the CLI?

File: lib/improver/tests/blending/weights/test_WeightsUtilities.py (hunk 15 above)
Comment (fionaRust) [reply lib/improver/tests/blending/weights/test_WeightsUtilities.py]: I've now changed this to add a new unit test for this rather than editing the existing one. It is still doing blending over forecast period, and the cube doesn't quite look like the input cubes we would expect in blending in reality, but the test does test the new functionality. Hopefully this is ok.

File: bin/improver-weighted-blending (hunk 16 above)
Comment (fionaRust) [reply bin/improver-weighted-blending]: I've now done this by adding the option to return a float for the number of grid squares to the function we already had to convert from m to grid lengths. I've then added a call to this function to the CLI.

File: tests/improver-weighted-blending/18-nowcast_cycle_blending.bats [hunk 20]
```diff
@@ -0,0 +1,53 @@
...
+# ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+# POSSIBILITY OF SUCH DAMAGE.
+
+. $IMPROVER_DIR/tests/lib/utils
+
+@test "weighted-blending model blending" {
+  improver_check_skip_acceptance
+  KGO="weighted_blending/spatial_weights/kgo/kgo_cycle.nc"
+
+  # Run weighted blending with linear weights for two input files and check it
```
Comment (MoseleyS) [inline tests/improver-weighted-blending/18-nowcast_cycle_blending.bats]: There are three input files, not two.

File: tests/improver-weighted-blending/19-spatial_model_blending.bats [hunk 21]
```diff
@@ -0,0 +1,51 @@
...
+# INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+# CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+# ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+# POSSIBILITY OF SUCH DAMAGE.
+
+. $IMPROVER_DIR/tests/lib/utils
+
+@test "weighted-blending model blending" {
+  improver_check_skip_acceptance
+  KGO="weighted_blending/spatial_weights/kgo/model_kgo.nc"
```
Comment (MoseleyS) [inline tests/improver-weighted-blending/19-spatial_model_blending.bats]: The kgo file names are inconsistent. The preceding test has kgo_cycle.nc and here it is model_kgo.nc. As they're in a kgo directory, perhaps the "kgo" bit isn't needed in the filenames and they would become consistently ordered.

**Comment by fionaRust** [comment]
@MoseleyS I've responded to your review comments now

File: lib/improver/blending/spatial_weights.py [hunk 22]
```diff
@@ -281,18 +357,17 @@ def process(self, cube_to_collapse, one_dimensional_weights_cube,
...
+            template_cube)
         weights_from_mask = self.smooth_initial_weights(
             weights_from_mask)
         final_weights = self.multiply_weights(
             weights_from_mask, one_dimensional_weights_cube,
             blend_coord)
         final_weights = self.normalised_masked_weights(
             final_weights, blend_coord)
-        # Check dimensions
-        final_weights = check_cube_coordinates(cube_to_collapse, final_weights)
```
Comment (fionaRust) [inline lib/improver/blending/spatial_weights.py]: Need to remove unused import of check_cube_coordinates
